File: core/src/main/java/com/github/fburato/highwheelmodules/core/externaladapters/GuavaGraphFactory.java
package com.github.fburato.highwheelmodules.core.externaladapters;

import com.github.fburato.highwheelmodules.model.modules.*;
import com.google.common.graph.NetworkBuilder;

import java.util.Optional;

public class GuavaGraphFactory implements ModuleGraphFactory {
    @Override
    public MetricModuleGraph<ModuleDependency> buildMetricModuleGraph() {
        return new GuavaModuleGraph(NetworkBuilder.directed().allowsSelfLoops(true).build());
    }

    @Override
    public GuavaEvidenceModuleGraph buildEvidenceModuleGraph(ModuleGraph<TrackingModuleDependency> trackingGraph,
            Optional<Integer> evidenceLimit) {
        return new GuavaEvidenceModuleGraph(trackingGraph, evidenceLimit);
    }

    @Override
    public GuavaTrackingModuleGraph buildTrackingModuleGraph() {
        return new GuavaTrackingModuleGraph(NetworkBuilder.directed().allowsSelfLoops(true).build());
    }
}

File: core/src/main/scala/com/github/fburato/highwheelmodules/core/externaladapters/GuavaGraphFactory.scala
package com.github.fburato.highwheelmodules.core.externaladapters

import com.github.fburato.highwheelmodules.model.modules._
import com.google.common.graph.NetworkBuilder

class GuavaGraphFactory extends ModuleGraphFactoryS {
  override def buildMetricModuleGraph: MetricModuleGraphS[ModuleDependencyS] =
    new GuavaModuleGraph(NetworkBuilder.directed.allowsSelfLoops(true).build)

  override def buildEvidenceModuleGraph(trackingGraph: ModuleGraphS[TrackingModuleDependencyS], evidenceLimit: Option[Int]): ModuleGraphS[EvidenceModuleDependencyS] =
    new GuavaEvidenceModuleGraph(trackingGraph, evidenceLimit)

  override def buildTrackingModuleGraph: ModuleGraphS[TrackingModuleDependencyS] =
    new GuavaTrackingModuleGraph(NetworkBuilder.directed.allowsSelfLoops(true).build)
}
